hail/src/main/scala/is/hail/shuffler/ShuffleClient.scala

```diff
@@ -0,0 +1,95 @@
+package is.hail.shuffler
+
+import org.apache.log4j.Logger;
+import is.hail.shuffler.ShuffleUtils._
+import is.hail.utils._
+import is.hail.annotations._
+import is.hail.expr.types.physical.{ PStruct, PType }
+import is.hail.expr.types.virtual.TStruct
+import is.hail.io.TypedCodecSpec
+import java.io._
+import java.net._
+import java.security.KeyStore;
+import java.util.UUID
+import java.util.concurrent.{ ConcurrentSkipListMap, Executors }
+import javax.net._
+import javax.net.ssl._
+import javax.security.cert.X509Certificate;
+import org.json4s.jackson.{ JsonMethods, Serialization }
+
+class ShuffleClient (
+  t: TStruct,
+  codecSpec: TypedCodecSpec,
+  key: Array[String],
+  ssl: SSLContext,
+  host: String,
+  port: Int
+) {
+  val log = Logger.getLogger(this.getClass.getName());
+
+  val sf = ssl.getSocketFactory()
+  val s = sf.createSocket(host, port)
+  val in = new DataInputStream(s.getInputStream())
+  val out = new DataOutputStream(s.getOutputStream())
+  log.info(s"CLNT connected to ${host}:${port}")
+
+  val keyedCodecSpec = new KeyedCodecSpec(t, codecSpec, key)
+  import keyedCodecSpec._
+
+  var uuid: String = null
+
+  def start(): Unit = {
+    log.info(s"CLNT start")
+    out.write(Wire.START)
+    Wire.writeTStruct(out, t)
+    Wire.writeTypedCodecSpec(out, codecSpec)
+    Wire.writeListOfStrings(out, key)
+    out.flush()
+    uuid = in.readUTF()
```

Comment:
I don't really like UUID for this.  I prefer pulling a block of bytes from the secure random number generator like we do generally in Python and using them as bytes (Array[Byte]) in this protocol.